iree/compiler/Conversion/LinalgToVector/VectorizeConv.cpp

```diff
@@ -0,0 +1,226 @@
+            return value.getSExtValue() != 1;
+          })) {
+        return failure();
+      }
+    }
+
+    auto filterViewOp = convOp.filter().getDefiningOp<SubViewOp>();
+    auto inputViewOp = convOp.input().getDefiningOp<SubViewOp>();
+    auto outputViewOp = convOp.output().getDefiningOp<SubViewOp>();
+    if (!filterViewOp || !inputViewOp || !outputViewOp) return failure();
+
+    // The filter/input/output view should have static sizes to vectorize.
+    if (!llvm::empty(filterViewOp.getDynamicSizes()) ||
+        !llvm::empty(inputViewOp.getDynamicSizes()) ||
+        !llvm::empty(outputViewOp.getDynamicSizes())) {
+      return failure();
+    }
+
+    // The output batch and height dimensions should be 1. If not, other
+    // patterns can generate parallel loops can distribute them.
+    if (outputViewOp.getStaticSize(0) != 1 ||
+        outputViewOp.getStaticSize(1) != 1) {
+      return failure();
+    }
+
+    // We addtionally expect the filter height/width dimensions are both 1 to
+    // simplify vectorization. Other patterns can generate loops to create 1x1
+    // filter subivews.
+    if (filterViewOp.getStaticSize(0) != 1 ||
+        filterViewOp.getStaticSize(1) != 1) {
+      return failure();
+    }
+
+    int64_t numInputChannels = filterViewOp.getStaticSize(2);
+    int64_t numOutputChannels = filterViewOp.getStaticSize(3);
+    if (numInputChannels != 4 || numOutputChannels % 4 != 0) return failure();
+
+    int64_t numOutputWidths = outputViewOp.getStaticSize(2);
+    int64_t widthStride = convOp.getStride(1);
+
+    // This invocation handles a batch of (numOutputWidths * numOutputChannels).
+    LLVM_DEBUG({
+      llvm::dbgs() << "# output width: " << numOutputWidths << "\n";
+      llvm::dbgs() << "# output channels: " << numOutputChannels << "\n";
+      llvm::dbgs() << "width stride: " << widthStride << "\n";
+    });
+
+    MLIRContext *context = convOp.getContext();
+    Location loc = convOp.getLoc();
+
+    auto vector1x4Type =
+        VectorType::get({1, 4}, filterViewOp.getType().getElementType());
+    Value zero = rewriter.create<ConstantIndexOp>(loc, 0);
+
+    // Load the entire filter subview. Both the height and width dimensions are
+    // 1; so we just need to loop over input and output channel dimensions.
+    SmallVector<SmallVector<Value, 1>, 4> filterVectors(numInputChannels);
+    for (int ic = 0; ic < numInputChannels; ++ic) {
+      auto &thisInputChannel = filterVectors[ic];
+      thisInputChannel.reserve(numOutputChannels / 4);
+      for (int oc = 0; oc < numOutputChannels / 4; ++oc) {
+        SmallVector<Value, 4> filterIndices(4, zero);
+        filterIndices[2] = rewriter.create<ConstantIndexOp>(loc, ic);
+        filterIndices[3] = rewriter.create<ConstantIndexOp>(loc, oc * 4);
+        thisInputChannel.push_back(rewriter.create<vector::TransferReadOp>(
+            loc, vector1x4Type, filterViewOp, filterIndices));
+      }
+    }
```

Comment:
So this already pre-unroll the vector operations. Would it be possible to generate large vector ops instead and use the unrolling patterns to do the unrolling separately? This would make things more flexible in the future and more generic in case we want to upstream this transformation? What do you think? Maybe I'm missing some complexity there.

Reply:
+1 to this suggestion. We should use unrolling to break the ops into smaller ops.

Reply:
Makes sense. Changed to load the filter as a whole and get slices from it so that we can rely on unrolling later.